pkg/eventpersister/persister/persister.go

```diff
@@ -198,39 +227,46 @@ func (p *Persister) batch() error {
 			if len(batch) < p.opts.flushSize {
 				continue
 			}
-			p.send(batch)
-			batch = make(map[string]*puller.Message)
-			timer.Reset(p.opts.flushInterval)
+			envEvents := p.extractEvents(batch)
+			// Update the feature flag last-used cache
+			p.cacheLastUsedInfoPerEnv(envEvents, envCache)
+			updateEvaluationCounter(envEvents)
 		case <-timer.C:
-			if len(batch) > 0 {
-				p.send(batch)
-				batch = make(map[string]*puller.Message)
-			}
-			timer.Reset(p.opts.flushInterval)
+			envEvents := p.extractEvents(batch)
+			// Update the feature flag last-used cache
+			p.cacheLastUsedInfoPerEnv(envEvents, envCache)
+			updateEvaluationCounter(envEvents)
+		case <-timerWriteCache.C:
+			p.logger.Debug("Write cache timer triggered")
+			// Write the feature flag last-used cache in the MySQL
+			p.writeEnvLastUsedInfo(envCache)
+			// Reset the cache and the timer
+			envCache = make(environmentLastUsedInfoCache)
+			timerWriteCache.Reset(p.opts.writeCacheInterval)
 		case <-p.ctx.Done():
+			// Because the feature flag last-used cache is not mandatory
+			// when the context is done, we don't try to write the cache
+			// to speed up the shutdown process
```

Comment:
Is it only performance reason to discard the cache?
IMO, last-used cache should be updated.

Reply:
The container has less than 10 seconds to shut down gracefully. If the PubSub message is not processed in time, it will be held for 10 minutes because it won't be Nack.

Reply:
I have refactored to update and write the cache when shutting down.
Btw, the current feature-recorder doesn't write the cache when shutting down.
https://github.com/bucketeer-io/bucketeer/blob/main/pkg/feature/recorder/recorder.go#L170